auditable/auditable.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package auditable
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+var (
+	apiPackagePath = "github.com/mattermost/mattermost-server/v6/api4"
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name: "auditable",
+	Doc:  "check if auditable interface is satisfied when passing structs to audit records",
+	Run:  run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	if pass.Pkg.Path() != apiPackagePath {
+		return nil, nil
+	}
+
+	for _, file := range pass.Files {
+		commentMap := make(map[token.Pos]*ast.CommentGroup)
+		for i := range file.Comments {
+			commentMap[file.Comments[i].Pos()] = file.Comments[i]
+		}
+
+		ast.Inspect(file, func(n ast.Node) bool {
+			expr, ok := n.(*ast.CallExpr)
+			if !ok {
+				return true
+			}
+
+			if expr.Fun == nil {
+				return true
+			}
+
+			fn, ok := expr.Fun.(*ast.SelectorExpr)
+			if !ok {
+				return true
+			}
+
+			if fn.Sel != nil && fn.Sel.Name != "AddEventParameter" {
+				return true
+			}
+
+			if len(expr.Args) < 2 {
+				return false
+			}
+
+			param, ok := expr.Args[1].(*ast.Ident)
+			if !ok {
+				return false
+			}
+
+			typ, ok := pass.TypesInfo.Types[param]
+			if !ok {
+				return false
+			}
+
+			// we check for the comment next to the whole expression, so we need to add 1 to the end position
+			// if we find a comment with the auditable:ignore tag, we skip the check for this node
+			if comment, ok := commentMap[expr.End()+1]; ok {
+				if strings.Contains(comment.Text(), "auditable:ignore") {
+					return false
+				}
+			}
+
+			if auditable, err := isAuditable(typ.Type); err == nil && !auditable {
+				pass.Reportf(param.Pos(), "%s is not auditable, but it is added to the audit record", typ.Type.String())
+			} else if err != nil {
+				pass.Reportf(param.Pos(), "error checking if %s is auditable: %v", typ.Type.String(), err)
+			}
+
+			return true
+		})
+	}
+
+	return nil, nil
+}
+
+func isAuditable(typ types.Type) (bool, error) {
+	var nt *types.Named
+	if _, ok := typ.(*types.Basic); ok {
+		return true, nil
+	}
+
+	switch v := typ; v.(type) {
+	case *types.Interface:
+		return false, nil
```

Comment:
I'm sure this is a non-issue in practice since unnamed interfaces are pretty rare, but shouldn't this case check the methods, too?

Reply:
It's a good call, but since the type of an `interface{}` can have any interface, it requires a dynamic check. We are actually working with static check tools so it's a hassle to figure out which interface does these implement or which methods does it have. The `*types.Interface` does not unwrap to it's actual interface with methods. At least I couldn't find a way to do that :/

Reply:
Yeah we can't check runtime types here obviously. I was thinking about the case where the input is a variable declared like this:

```go
	var foo interface {
		Auditable() map[string]interface{}
	}
```

Reply:
Okay, I got your point. Yes, then we can resolve it for sure, we can check the methods. I'll add it.

Reply:
Done.